workflow/src/test/java/bio/terra/tanagra/workflow/GraphUtilsTest.java

```diff
@@ -0,0 +1,113 @@
+package bio.terra.tanagra.workflow;
+
+import com.google.common.collect.Multimap;
+import com.google.common.collect.MultimapBuilder;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.Distinct;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class GraphUtilsTest {
+
+  @Rule public final transient TestPipeline pipeline = TestPipeline.create();
+
+  /**
+   * Runs a test pipeline with the input edges and {@link GraphUtils#transitiveClosure}, asserting
+   * the expected transitive edges are produced.
+   */
+  <T> void runTransitiveClosureAndAssert(
+      Multimap<T, T> edges, Multimap<T, T> expectedEdges, int maxPathLength) {
+    PCollection<KV<T, T>> transitiveEdges =
+        GraphUtils.transitiveClosure(pipeline.apply(Create.of(convertToKvs(edges))), maxPathLength)
+            .apply(Distinct.create()); // The transitiveClosure may output duplicates.
+    PAssert.that(transitiveEdges).containsInAnyOrder(convertToKvs(expectedEdges));
+    pipeline.run().waitUntilFinish();
+  }
+
+  /** Convert a multimap to an equivalent list of KVs. */
+  private static <T> List<KV<T, T>> convertToKvs(Multimap<T, T> multimap) {
+    return multimap.entries().stream()
+        .map(entry -> KV.of(entry.getKey(), entry.getValue()))
+        .collect(Collectors.toList());
+  }
+
+  @Test
```

Comment:
opt: may be worth including some smaller edge cases for the logarithmic approach, e.g. single-edge graph, with a maxPathLen of 1

Reply:
Good idea, added.